### PR TITLE
OCLOMRS-207: Add CIEL concepts in the add Bulk concept page

### PR DESCRIPTION
**Pull request by kmunene**
# JIRA TICKET NAME:
[Add CIEL concepts in the add Bulk concept page](https://issues.openmrs.org/browse/OCLOMRS-207)

# Summary:
Currently, a user cannot add bulk CIEL concepts. This issue will enable a user to add bulk CIEL concepts in the add Bulk page.

### CODE REVIEW

File: src/components/bulkConcepts/addBulkConcepts.jsx [hunk 1]
```diff
@@ -1,37 +1,88 @@
...
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      currentPage: 1,
+      cielConcepts: [],
+    };
+  }
+
   handleAddAll=() => {
```
Comment (dondrzzy) [inline src/components/bulkConcepts/addBulkConcepts.jsx]: Please space this right. `All = () => {}`

File: src/components/bulkConcepts/addBulkConcepts.jsx [hunk 2]
```diff
@@ -1,37 +1,88 @@
...
-    const expressions = this.props.cielConcepts.map(concept => concept.url);
-    this.props.addExistingBulkConcepts({ data: { expressions } });
+    const { type, typeName, collectionName } = this.props.match.params;
```
Comment (dondrzzy) [inline src/components/bulkConcepts/addBulkConcepts.jsx]: Hey @kmunene, Is your ESlint configured well. Please use destructuring props assignment.

File: src/components/bulkConcepts/addBulkConcepts.jsx [hunk 3]
```diff
@@ -1,37 +1,88 @@
...
+    const url = `${type}/${typeName}/collections/${collectionName}/references/`;
+    const data = this.state.cielConcepts;
```
Comment (dondrzzy) [inline src/components/bulkConcepts/addBulkConcepts.jsx]: here too

File: src/components/bulkConcepts/addBulkConcepts.jsx [hunk 4]
```diff
@@ -1,37 +1,88 @@
...
+    this.props.addExistingBulkConcepts({ url, data: { data: { expressions: data } } });
+    setTimeout(() => {
```
Comment (dondrzzy) [inline src/components/bulkConcepts/addBulkConcepts.jsx]: Why did you add a 2 seconds delay?

File: src/components/bulkConcepts/addBulkConcepts.jsx [hunk 5]
```diff
@@ -1,37 +1,88 @@
...
+      window.history.back();
+    }, 2000);
   }
+
   handleCielClick=() => {
     this.props.fetchCielConcepts();
   }
+
+  handleBack=() => {
```
Comment (dondrzzy) [inline src/components/bulkConcepts/addBulkConcepts.jsx]: Proper spacing required.

File: src/components/bulkConcepts/addBulkConcepts.jsx [hunk 6]
```diff
@@ -1,37 +1,88 @@
...
+      });
+    } else {
+      const index = this.state.cielConcepts.indexOf(value);
+      this.setState({
+        cielConcepts: this.state.cielConcepts.filter((_, i) => i !== index),
+      });
+    }
+  }
+
+  handleClick = (e) => {
```
Comment (dondrzzy) [inline src/components/bulkConcepts/addBulkConcepts.jsx]: Is there one click event on this component. More description of the event handler would seem better. What do you think?

File: src/tests/bulkConcepts/addBulkConcepts.test.jsx [hunk 7]
```diff
@@ -51,4 +52,52 @@ describe('Add Bulk Concepts', () => {
...
+
+  it('it calls hanldleBack', () => {
+    const props = {
+      fetchCielConcepts: jest.fn(),
+      addExistingBulkConcepts: jest.fn(),
+      cielConcepts: multipleConceptsMockStore.concepts.dictionaryConcepts,
+      isFetching: true,
+      match: { params: { type: 'users', typeName: 'mochu', collectionName: 'andela' } },
+    };
+    const wrapper = mount(<MemoryRouter>
```
Comment (dondrzzy) [inline src/tests/bulkConcepts/addBulkConcepts.test.jsx]: You have repeated some code in the above test. Please create a beforeEach and place the repeated code in there, implement DRY programming.

File: src/tests/bulkConcepts/addBulkConcepts.test.jsx [hunk 8]
```diff
@@ -51,4 +52,52 @@ describe('Add Bulk Concepts', () => {
...
+      <Provider store={store}>
+        <AddBulkConcepts {...props} />
+      </Provider>
+    </MemoryRouter>);
+    const spy = jest.spyOn(wrapper.find('AddBulkConcepts').instance(), 'handleBack');
+    wrapper.instance().forceUpdate();
+    wrapper.find('.btn.btn-secondary').at(0).simulate('click');
```
Comment (dondrzzy) [inline src/tests/bulkConcepts/addBulkConcepts.test.jsx]: I don't think this method of finding elements is the best. More so when you are using bootstrap classes that any other developer could add to the component. Maybe assign a unique class if you are certain the back buttons are many. If it's one button. Give it an ID and use that in your find. How about that?

File: src/components/bulkConcepts/addBulkConcepts.jsx (hunk 1 above)
Comment (alexmochu) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: +1 on this

**Comment by dkayiwa** [comment]
@kmunene it is now two weeks and this is still open. Why?

**Comment by dkayiwa** [comment]
@kmunene why is this still open?

**Comment by dkayiwa** [comment]
@kmunene did you see my question above?

**Comment by kmunene** [comment]
@dkayiwa. I have now seen it. I am working on the requested changes and a bug I encountered. Will be done with this changes before tomorrow morning.

**Comment by dkayiwa** [comment]
@kmunene am still struggling to figure out why you do not respond to my comments.

**Comment by kmunene** [comment]
If the comments were made before Friday, I had not turned on my Github notifications. I now have done this and can immediately respond to any comments made on the PR.

**Comment by dkayiwa** [comment]
The second comment was not made before Friday

**Comment by kmunene** [comment]
Then, I apologize for not replying to it at the right time. However, I was not ignoring your comments.

File: src/components/bulkConcepts/addBulkConcepts.jsx (hunk 1 above)
Comment (kmunene) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: I have fixed this.

File: src/components/bulkConcepts/addBulkConcepts.jsx (hunk 2 above)
Comment (kmunene) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: I have fixed this.

File: src/components/bulkConcepts/addBulkConcepts.jsx (hunk 3 above)
Comment (kmunene) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: I have fixed this.

File: src/components/bulkConcepts/addBulkConcepts.jsx (hunk 4 above)
Comment (kmunene) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: The 2 second time delay is for transitioning from one page to another. I felt that it should not be sudden.

File: src/components/bulkConcepts/addBulkConcepts.jsx (hunk 5 above)
Comment (kmunene) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: I have fixed this.

File: src/components/bulkConcepts/addBulkConcepts.jsx (hunk 6 above)
Comment (kmunene) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: That's a good Idea. I will fix this.

File: src/tests/bulkConcepts/addBulkConcepts.test.jsx (hunk 7 above)
Comment (kmunene) [reply src/tests/bulkConcepts/addBulkConcepts.test.jsx]: Sure. Thank you for noticing that.

File: src/tests/bulkConcepts/addBulkConcepts.test.jsx (hunk 8 above)
Comment (kmunene) [reply src/tests/bulkConcepts/addBulkConcepts.test.jsx]: I agree with you on this one. Will make the necessary changes.

File: src/components/bulkConcepts/addBulkConcepts.jsx [hunk 9]
```diff
@@ -1,36 +1,95 @@
...
+        params: {
+          type, typeName, collectionName,
+        },
+      },
+    } = this.props;
+    const url = `${type}/${typeName}/collections/${collectionName}/references/`;
+    const { cielConcepts } = this.state;
+    this.props.addExistingBulkConcepts({ url, data: { data: { expressions: cielConcepts } } });
+    setTimeout(() => {
+      window.history.back();
```
Comment (dondrzzy) [inline src/components/bulkConcepts/addBulkConcepts.jsx]: You defined a function that does this. I think you should use it here.

File: src/components/bulkConcepts/addBulkConcepts.jsx (hunk 4 above)
Comment (dondrzzy) [reply src/components/bulkConcepts/addBulkConcepts.jsx]: What is wrong with it being immediate? You are not depending on the result response from that API call and neither are you doing anything inside the setTimeout function so maybe it should not be there.

File: src/components/bulkConcepts/addBulkConcepts.jsx [hunk 10]
```diff
@@ -1,36 +1,95 @@
...
+  handleSelect = (e) => {
+    e.preventDefault();
+    const { value, checked } = e.target;
+
+    if (checked) {
+      this.setState({
+        cielConcepts: [value, ...this.state.cielConcepts],
+      });
+    } else {
+      const index = this.state.cielConcepts.indexOf(value);
```
Comment (dondrzzy) [inline src/components/bulkConcepts/addBulkConcepts.jsx]: Use destructuring props assignment.

File: src/tests/bulkConcepts/addBulkConcepts.test.jsx [hunk 11]
```diff
@@ -51,4 +53,138 @@ describe('Add Bulk Concepts', () => {
...
+    };
+
+    const spy = jest.spyOn(wrapper.find('AddBulkConcepts').instance(), 'handleSelect');
+    wrapper.instance().forceUpdate();
+    wrapper.find('#ciel').at(0).simulate('click');
+    wrapper.find('.table-check').at(0).simulate('change', event);
+    expect(spy).toHaveBeenCalledTimes(1);
+  });
+
+  it('it calls hanldleSelect and remove a concept', () => {
```
Comment (dondrzzy) [inline src/tests/bulkConcepts/addBulkConcepts.test.jsx]: Does this sound right while you read it?

File: src/tests/bulkConcepts/addBulkConcepts.test.jsx [hunk 12]
```diff
@@ -51,4 +53,138 @@ describe('Add Bulk Concepts', () => {
...
+      <Provider store={store}>
+        <AddBulkConcepts {...props} />
+      </Provider>
+    </MemoryRouter>);
+    const spy = jest.spyOn(wrapper.find('AddBulkConcepts').instance(), 'handlePaginationClick');
+    wrapper.instance().forceUpdate();
+    wrapper.find('#ciel').at(0).simulate('click');
+    wrapper.find('.fas.fa-angle-double-right.nxt').at(0).simulate('click');
+    expect(spy).toHaveBeenCalledTimes(1);
+  });
```
Comment (dondrzzy) [inline src/tests/bulkConcepts/addBulkConcepts.test.jsx]: I think you should initiate a beforeEach function in this file because there is a lot of repeated code as previously stated. However only use it in the test code blocks you have added.
